TSS-3221: Write the script queue

r-owen opened the pull request.

tribeiro commented:
When adding an item to the queue if, `location` and `locationSalIndex` are left with the default set of values (`0`), I get a command rejection with `Failed: Unknown location 0`. I think that it would be good if the default values implemented the default behavior, which is append to the queue. Right?

r-owen commented:
@tribeiro you raise an excellent point. However, I'm afraid it's not practical because the value has to be an enum value and enum values in our XML files have to start at 1. I would gladly have made the default "last" if I could set my own enum values. (Yes I could probably cheat and accept 0 even though it is not a valid enum value, but I am very reluctant to do that).

I don't think this will cause much pain in real life because the telescope should have a GUI for managing the queue.

timj commented:
You are saying that there is no easy way to add to the back of the queue without knowing how long the queue already is? I don't like that at all.  I had envisaged a command line tool for adding things to the queue. I guess that the command line tool could be clever enough to have to ask first for where the end is. Is the scheduler always having to track the current end position? (maybe it is).

tribeiro commented:
No @timj , that index is just an enumeration for `Front`, `Back`, `Before` or `After`.... You only need to specify position if you select `Before` or `After`.... 

If you don't fill that information the command gets rejected... Also, the "default" behavior is achieved with `Back`, so `location = 2` is how you get add to back... :/

tribeiro commented:
One quick question. So, right now the queue has a `pause` command but no `stop`, right? Is the feature to pause the queue and interrupt a running script (e.g. `stop`) going to be added in the future? I imagine a *running script* must have a `stop` method that has knowledge on how to gracefully interrupt an executing task, is that right?

tribeiro commented:
So, I understand the queue status is either running or paused. But shouldn't it be part of the queue logic to be able to not only pause itself but also interrupt a running script? Even if the actual procedure is to pause the queue and send a stop command to a running script, I think it would be good to be able to achieve that behavior with a single command to the queue. 

Regardless of the queue being the entity that runs the scripts, don't you think there should be a way for the queue to send a stop command to a script? Otherwise I think the interaction becomes quite odd. You add scripts to the queue to run but when you want to stop them, you have to pause the queue and send a command to the script to stop. I'd rather just send a stop command to the queue and it takes cares of it.

timj commented:
In LTS-836 initial design (a push queue) if the observation stopped for some reason other than normal completion the queue would automatically go into pause state.  You would ask the OCS to stop the currently running observation and the queue would notice this.  If the queue is running the observations then the queue should have some way of noticing that an observation has been stopped abnormally. Who receives that command depends on what is running the observation.

tribeiro commented:
@r-owen , when trying to send a `pause` command to the queue I get `ack=-302`, `error=1` and  `result=Failed: pause not allowed in state 5`.  I think that's because the queue don't have an enable state and you do https://github.com/lsst-ts/ts_scriptqueue/blob/a06d8e87063617a79837a42200f02ccff8122338/python/ts_scriptqueue/script_queue.py#L135

r-owen commented:
@tribeiro To stop the current job use the `remove` command. You can use this to remove the running script or any queued script. I feel this is safer than a command to stop the current job because that can easily get into a race condition where you are trying to stop one job, but it finishes and so you end up stopping the next job instead.

Remove cancels the job but does not pause the queue. I would suggest that the user pause the queue and then kill the job if that's what's wanted. However, if we want one command that does both then I could add a "doPause" flag to the "remove" command.

r-owen commented:
@timj Your comments made me realize I had left something out of the current code: if the current job fails then queue is *not* automatically paused. It would be fairly easy to add this feature. The question is whether to do it on this ticket or merge this and add it later.

r-owen commented:
@tribeiro I'm puzzled by your statement "when trying to send a pause command.... I think that's because the queue don't have an enable state...". The ScriptQueue has all the normal CSC states and it must be enabled for any of queue commands to run (add, move, remove, pause, etc.). It's Scripts that don't have normal CSC states because Scripts aren't CSCs. Does this explain what's going on, or are you still having a problem?

tribeiro commented:
@r-owen , now I am puzzled... I start the queue with the script provided here `bin.src/run_script_queue.py`. Once I do that, and without enabling the queue, I can send an `add` command to the queue and it will execute the scripts. And if I send `pause` or `resume` I get the error specified above. That is, a `-302` ack with the message saying that `Failed: pause not allowed in state 5`.

Also, note that if the intended behavior is for the queue to only accept `add` commands when it is enabled, then, I think there is a `self.assert_enabled("add")` missing here: https://github.com/lsst-ts/ts_scriptqueue/blob/a06d8e87063617a79837a42200f02ccff8122338/python/ts_scriptqueue/script_queue.py#L154

If I comment https://github.com/lsst-ts/ts_scriptqueue/blob/a06d8e87063617a79837a42200f02ccff8122338/python/ts_scriptqueue/script_queue.py#L135 and https://github.com/lsst-ts/ts_scriptqueue/blob/a06d8e87063617a79837a42200f02ccff8122338/python/ts_scriptqueue/script_queue.py#L146 then `pause` and `resume` works fine...

tribeiro commented:
Also, another question, how can I make a script fail? I'm forcing my script to trow an exception in the `run` method but the script is still marked as `completed`.

r-owen commented:
@tribeiro If a script throws an exception in `run` it should definitely be marked as failing. It looks like I forgot to call sys.exit(nonzero) on error. I'll fix that.

tribeiro commented:
Also, I don't think it is marking scripts as `Running`... I had a long script put on the queue and it stayed as `Configured` while it was running...

tribeiro commented:
@r-owen,  a minor note. 

I think the directory tree of the repository is not following the standard convention. In case of a package like `ts_scriptqueue` I think we expect to get something like `python/ts/scriptqueue` instead of `python/ts_scriptqueue`. 

I don't think anyone will be importing things from this package but it may be something to consider...

tribeiro commented on python/ts_scriptqueue/script_info.py:
```diff
@@ -0,0 +1,254 @@
+        if self.process.returncode is not None:
+            return False
+
+        self.process.terminate()
```
I think that before sending a `terminate` to the subprocess, we should give the script the chance to do some cleanup. Maybe there should be a command that we could send to the script before terminating. 

For instance, scripts to take long exposures could send a stop exposure command to the camera before terminating.

r-owen replied:
You raise a good point. Scripts can be gently shut down by sending "stop" but the script queue should probably offer a way to send that command. How about if I replace the "remove" ScriptQueue command with "stop" and "terminate"? Users should generally try "stop" before trying "terminate".

tribeiro replied:
yes, I like `stop` and `terminate` commands.

r-owen replied:
I have implemented `stop` and `terminate` commands. If the script is running then `stop` will give the script a chance to clean up, but if that takes too long or the script is not running then it will terminate the script. So `stop` is gentler than `terminate` but should be just as reliable.

r-owen commented:
@tribeiro regarding add working when not enabled: you are right. I missed a lot of "assert_enabled" calls. I suspect salobj should make all commands require enabled state except for the standard state transition commands. I'm working on cleaning this up now. It affects a lot of code in small ways (for instance the constructor cannot call do_showAvailable scripts and the model should know if the ScriptQueue is enabled).

r-owen commented:
@tribeiro I think I have fully fixed issues with handling enabled and non-enabled states. I have intentionally allowed "pause" to be used in any state, but the other non-standard commands are all rejected and the queue will not run a new script unless the CSC is enabled. This is in a new commit to ts_scriptqueue and another commit to ts_xml (to add the `enabled` field to the `queue` event).

r-owen commented:
> I think the directory tree of the repository is not following the standard convention. In case of a package like `ts_scriptqueue` I think we expect to get something like `python/ts/scriptqueue` instead of `python/ts_scriptqueue`.

I've been wondering about this. I'd be happy to use ts/ instead of ts_ but it means that users *must* have the DM `utils` package, which adds a dependency. If I don't do that then there are no DM dependencies at all (though sconsUtils is useful for running unit tests).

tribeiro commented:
> > I think the directory tree of the repository is not following the standard convention. In case of a package like `ts_scriptqueue` I think we expect to get something like `python/ts/scriptqueue` instead of `python/ts_scriptqueue`.
> 
> I've been wondering about this. I'd be happy to use ts/ instead of ts_ but it means that users _must_ have the DM `utils` package, which adds a dependency. If I don't do that then there are no DM dependencies at all (though sconsUtils is useful for running unit tests).

I personally prefer `ts/` instead of `ts_` and I think it would worth having DM `utils` installed if that's what it takes. But I can leave with it either way.

tribeiro commented:
Mentioning @taxelrod to make sure he follows the discussion here.

tribeiro commented:
@r-owen and @timj some other detail. Isn't the Queue supposed to be indexed as well? Aren't we supposed to have a queue for auxtel and one for main telescope?

tribeiro commented:
@r-owen another point that concerns me is the time required to initialize a script. It adds quite an overhead (of almost 10 seconds) to run a script. We spoke about this already, I just want to make sure it will be possible to fix this on sal in the future.

r-owen commented:
Regarding the ScriptQueue not being indexed: we can certainly do that if we have to, but I figured we'd start without it and try to run both telescopes off of a single queue. If we do need two separate queues we'll also need to separate the ranges of indices for the scripts so there is no overlap. This will be a simple enough change.

tribeiro commented:
> Regarding the ScriptQueue not being indexed: we can certainly do that if we have to, but I figured we'd start without it and try to run both telescopes off of a single queue. If we do need two separate queues we'll also need to separate the ranges of indices for the scripts so there is no overlap. This will be a simple enough change.

I think that, to accomplish that, we will need to be able to run more than one script concurrently. Personally, I like the idea, but it may add a lot of complexity to the queue.

r-owen commented:
Regarding the `remove` command: I have replaced that with `stop` and `terminate` commands. This required an update to `ts_xml` and I took the opportunity to rebase that to the latest `develop` branch. I also think I fixed Scripts to return an exit code of 1 on failure (but 0 on success or if stopped). But I need to add a test for that and perhaps tweak up that code.

r-owen commented:
I just pushed a few more changes:
- I fixed `BaseScript.main` to halt with a suitable return code: 0 for DONE or STOPPED, 1 for FAILURE, 2 if the final state is something else (which should never happen).
- The script queue will now pause if a script fails. This leaves the failed script as the current script, in hopes this will make it easier to see why the queue paused. Issuing "resume" will move the failed script to history and resume normal operation.

r-owen commented:
@tribeiro @taxelrod Regarding the time to load a script: yes it's slow. This is TSS-3191. Personally this is my highest priority for a SAL improvement but I'm afraid it's up to Dave and management to make this happen.

tribeiro commented on python/ts/scriptqueue/script_queue.py:
```diff
@@ -0,0 +1,287 @@
+        """
+        self.assert_enabled("add")
+        script_info = ScriptInfo(
+            index=self.model.next_sal_index,
```
I think it would be good if `do_add` returned the script index. This would give the user a direct way to check the state of a requested script.

r-owen replied:
The only thing do_add can return to a remote user is the command ID. That is why the `script` event includes command ID -- it makes a fairly simple and completely reliable way to associate a particular script with a particular add command.

tribeiro replied:
Got it!

r-owen commented:
OK. I am convinced. I will make ts_scriptqueue indexed. I may have a shot at that today or may decide to make it a separate ticket. It should be a fairly simple change. My thought is that the index can be 1 or 2 (using enums in XML, which must start at one) and that the index will control the range of script indices, something like 1,000-99,999 for index 1 = main telescope and 100,000-199,999 for aux tel.). Indices will wrap around when we run out. The max index is 2,147,483,647 so we could provide larger ranges, but I would rather use shorter indices. And as long as wraparound takes at least a few days, we should be fine.

r-owen commented:
I just pushed changes to `ts_scriptqueue` and `tx_xml` that make ScriptQueue an indexed SAL component. Use 1 for the main telescope and 2 for AuxTel. Scripts SAL indices will start out 10,000 - 19,000 for Main and 20,000 - 29,000 for AuxTel but when Dave Mills fixes a small ts_sal bug, and that version is more widely used, I would like to increase the range by a factor of 10 to completely eliminate any worries about duplicate indices for scripts that run within a few days of each other.

tribeiro commented:
I'm still seeing scripts that failed as being completed.

tribeiro commented:
@r-owen The same also happens for scripts that are stopped and terminated... They all show as `Complete` (e.g. `processState = 4`)...

r-owen commented:
Scripts that are stopped have a return code of 0 by design because it has not failed, it has done what it was told to do. I have mixed feelings about that, but I think it's more sensible than a non-zero exit code. So a running script that is stopped by the queue with the "stop" command falls into that category. A failed script returns a non-zero code and this is unit tested. I'll look at what the script queue is doing with that..

r-owen commented:
I tracked down the problem with terminated scripts: calling terminate() on an asyncio.Process is supposed to result in a negative return code, but instead the code is 0. I'll have to think about that one. I can't see anything wrong reported state if a script fails; I added a test for this and it works.

r-owen commented:
I have pushed a fix for detecting terminated scripts. I also modified the code to prevent tests from being run in multiple processes to make them more reliable. It slows down things and ruins the coverage report, but I think the trade-off is acceptable for now. I hope TSS-3191 will eliminate the need but I don't really know why it's happening; it appears to be some kind of resource starvation.

tribeiro commented:
> Scripts that are stopped have a return code of 0 by design because it has not failed, it has done what it was told to do. I have mixed feelings about that, but I think it's more sensible than a non-zero exit code. So a running script that is stopped by the queue with the "stop" command falls into that category. A failed script returns a non-zero code and this is unit tested. I'll look at what the script queue is doing with that..

So maybe it is the case to add a new field that specifies how the execution ended. Something like  `successful`, `stoped` or `terminated`. Specially if we are planning to allow requeue of process, it helps a lot to be able to look to the past queue and see which one was stoped and so on...

r-owen commented:
@tribeiro I agree. I've been reluctant to add too much of the Script API to ScriptQueue (since it's duplication) but I think this is one more piece that should be included. It will probably take me a few days. I am inclined to do that work on this ticket branch, unless you need a working version sooner, so as to reduce churn -- especially since it will require a modification to the XML files.

I am starting to wonder if we should also support setting checkpoints and pausing scripts via the ScriptQueue, thus making the entire Script API usable via the ScriptQueue.

tribeiro commented:
I think it would be ok to add the work to this ticket. We do want to start using the ScriptQueue soon but I think we can wait until  mid-/end-next week.

r-owen commented:
OK. I added the scriptState field to evt_script and simplified the processState enum slightly. There is still some overlap due to differences between what the ScriptQueue has asked the script to do and what the Script is reporting that it is doing, but I think it will be better to leave that, as the code is simpler and it's easier to see what's going on if something misbehaves. For the same reason many state changes will produce pairs of script events, configuring a script, running a script and terminating a script. I have not thought of a nice way to prevent this and it does reflect reality.

Finally, I updated the ts_scriptqueue package to use salobj 2.3.

To run this new version you will need to build a new SALPY_ScriptQueue from the latest ts_xml tickets/TSS-3221 branch and use salobj 2.3.0 or later.

r-owen commented:
Still on my "to do" list: support stopping or terminating more than one script in a single call, and perhaps supporting "purge all queued scripts". These are easy to add. The question is: should I do it on this ticket branch or wait?

tribeiro commented:
I would suggest doing it later.
